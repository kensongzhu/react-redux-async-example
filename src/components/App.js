import React from "react";
import Picker from "./Picker";
import Posts from "./Posts";


export default class App extends React.Component {

	render() {
		const {
			selectedSubreddit,
			posts,
			isFetching,
			lastUpdated,
			handleChange,
			handleRefreshClick
		} = this.props;
		const isEmpty = posts.length === 0;

		return (
			<div>
				<Picker
					value={selectedSubreddit}
					onChange={handleChange}
					options={['reactjs', 'frontend']}
				/>
				<p>
					{lastUpdated &&
					<span>
						Last update at {new Date(lastUpdated).toLocaleDateString()}.{' '}
					</span>
					}
					{!isFetching && <button onClick={handleRefreshClick}>Refresh</button>}
				</p>
				{isEmpty
					? (isFetching ? <h2>Loading...</h2> : <h2>Empty.</h2>)
					: <div style={{opacity: isFetching ? 0.5 : 1}}>
						<Posts posts={posts}/>
					</div>
				}
			</div>
		)
	}
}