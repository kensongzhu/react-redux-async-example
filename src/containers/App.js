import React from "react";
import Picker from "../components/Picker";
import Posts from "../components/Posts";
import {connect} from "react-redux";
import {fetchPostsIfNeeded, invalidateSubreddit, selectSubreddit} from "../actions";

class App extends React.Component {

	componentDidMount() {
		const {dispatch, selectedSubreddit} = this.props;
		dispatch(fetchPostsIfNeeded(selectedSubreddit));
	}

	componentDidUpdate(prevProps, prevState, snapshot) {
		if (prevProps.selectedSubreddit !== this.props.selectedSubreddit) {
			const {dispatch, selectedSubreddit} = this.props;
			dispatch(fetchPostsIfNeeded(selectedSubreddit));
		}
	}

	handleChange = nextSubreddit => {
		this.props.dispatch(selectSubreddit(nextSubreddit));
	};

	handleRefreshClick = e => {
		e.preventDefault();

		const {dispatch, selectedSubreddit} = this.props;
		// why invalidate?
		dispatch(invalidateSubreddit(selectedSubreddit));
		dispatch(fetchPostsIfNeeded(selectedSubreddit));
	};

	render() {
		const {selectedSubreddit, posts, isFetching, lastUpdated} = this.props;
		const isEmpty = posts.length === 0;

		return (
			<div>
				<Picker
					value={selectedSubreddit}
					onChange={this.handleChange}
					options={['reactjs', 'frontend']}
				/>
				<p>
					{lastUpdated &&
					<span>
						Last update at {new Date(lastUpdated).toLocaleDateString()}.{' '}
					</span>
					}.

					{!isFetching && <button onClick={this.handleRefreshClick}>Refresh</button>}
				</p>
				{isEmpty
					? (isFetching ? <h2>Loading...</h2> : <h2>Empty.</h2>)
					: <div style={{opacity: isFetching ? 0.5 : 1}}>
						<Posts posts={posts}/>
					</div>
				}
			</div>
		)
	}
}

const mapStateToProps = state => {
	const {selectedSubreddit, postsBySubreddit} = state;
	const {isFetching, lastUpdated, items: posts} = postsBySubreddit[selectedSubreddit] || {isFetching: true, items: []}
	return {
		selectedSubreddit,
		posts,
		isFetching,
		lastUpdated
	}
};


export default connect(mapStateToProps)(App)