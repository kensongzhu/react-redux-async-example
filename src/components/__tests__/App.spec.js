import React from "react";
import App from "../App";
import {shallow} from "enzyme";

const setup = (setupProps = {}) => {
    const defaultProps = {
        selectedSubreddit: 'frontend',
        isFetching: false,
        posts: [],
        handleChange: jest.fn(),
        handleRefreshClick: jest.fn()
    };
    const props = {...defaultProps, ...setupProps};
    const wrapper = shallow(<App {...props} />);

    return {
        props,
        wrapper
    }
};

describe('App Component', () => {


    test("should render without crashing", () => {
        const {wrapper} = setup();
        expect(wrapper).toMatchSnapshot();
    });

    test("should render last updated and the posts", () => {
        const {wrapper} = setup({
            lastUpdated: 1482363367071,
            posts: [
                {title: 'frontend post 1'},
                {title: 'frontend post 2'},
                {title: 'frontend post 3'},
            ]
        });
        expect(wrapper).toMatchSnapshot();
    })
});