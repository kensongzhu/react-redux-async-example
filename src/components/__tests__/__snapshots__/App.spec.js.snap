// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render last updated and the posts 1`] = `
<div>
  <Picker
    onChange={[MockFunction]}
    options={
      Array [
        "reactjs",
        "frontend",
      ]
    }
    value="frontend"
  />
  <p>
    <span>
      Last update at 
      12/22/2016
      .
       
    </span>
    <button
      onClick={[MockFunction]}
    >
      Refresh
    </button>
  </p>
  <div
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Posts
      posts={
        Array [
          Object {
            "title": "frontend post 1",
          },
          Object {
            "title": "frontend post 2",
          },
          Object {
            "title": "frontend post 3",
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`App Component should render without crashing 1`] = `
<div>
  <Picker
    onChange={[MockFunction]}
    options={
      Array [
        "reactjs",
        "frontend",
      ]
    }
    value="frontend"
  />
  <p>
    <button
      onClick={[MockFunction]}
    >
      Refresh
    </button>
  </p>
  <h2>
    Empty.
  </h2>
</div>
`;
